### PR TITLE
Return redirection for symlinks

### PR DESCRIPTION
**Pull request by piranna**
This fixes issue #86

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -12,31 +12,38 @@
...
+var fresh       = require('fresh')
+var mime        = require('mime')
+var ms          = require('ms');
+var onFinished  = require('on-finished')
+var parseRange  = require('range-parser')
+var statuses    = require('statuses')
+
+var debug     = require('debug')('send')
 var deprecate = require('depd')('send')
-var destroy = require('destroy')
```
Comment (dougwilson) [inline index.js]: Please undo all these changes.

**Comment by dougwilson** [comment]
Hi! Thanks for the pull request! Please undo all the style changes, mainly because they are unwelcome (we weren't asking for style changes) and it makes it really hard for me to understand what changed from looking at the diff.

Also, please add documentation and tests for the new feature so we can accept it :)

File: index.js (hunk 1 above)
Comment (piranna) [inline index.js]: I just put the requires sorted alphabetically and grouped by build-ins, third-parties and local modules, as it's usually accepted in the Node.js community. Also I've made the requires style uniform, instead of some of them using its own `var` statement and others a common one, and some requires that were hidden between the variables declarations. Do you still want these clean-up changes to be undone?

File: index.js [hunk 2]
```diff
@@ -57,6 +64,35 @@ module.exports.mime = mime
 var listenerCount = EventEmitter.listenerCount
   || function(emitter, type){ return emitter.listeners(type).length; };
 
+
+function responseStatus(res, code, msg)
+{
+  if(msg == null)
```
Comment (dougwilson) [inline index.js]: Please match existing styles, for example the bracket style really stands out as misplaced.

File: index.js [hunk 3]
```diff
@@ -396,28 +427,42 @@ SendStream.prototype.isRangeFresh = function isRangeFresh(){
  * @private
  */
 
-SendStream.prototype.redirect = function redirect(path) {
+SendStream.prototype.redirectDirectory = function redirectDirectory(path) {
```
Comment (dougwilson) [inline index.js]: Changing this public method name is a breaking change, meaning accepting this would require a major version bump. There is no timeline for when a major will happen due to Express support, if that is OK.

File: index.js (hunk 1 above)
Comment (dougwilson) [inline index.js]: > Do you still want these clean-up changes to be undone?

Yes. It is extremely important to us that git blame stays intact, thus why there is a mixed style.

File: index.js [hunk 4]
```diff
@@ -396,28 +427,42 @@ SendStream.prototype.isRangeFresh = function isRangeFresh(){
...
   if (listenerCount(this, 'directory') !== 0) {
-    this.emit('directory')
-    return
+    return this.emit('directory')
```
Comment (dougwilson) [inline index.js]: Please leave the return on it's own line. We want to return undefined, not the event emitter instance.

**Comment by dougwilson** [comment]
Hi! Your original suggested I agreeded to was "Add an option to show the symbolic links as 301 redirections to the file location instead of return the file content.". I don't see an option to enable the feature. If this always enabled? If not, what is the option? We need an option to enable/disable this feature. If you want it accepted any time soon, you should probably default it to disabled so it's backwards compatible.

**Comment by dougwilson** [comment]
Oh, sorry, looks like you added a test from when I looked at it hours ago, haha. Please try to add more tests, including non happy paths. A new feature should have many tests, not just one, as there are probably edge cases you need to test for, for example, what about folder junctions vs folder symlinks? I wouldn't expect a junction to redirect, is that the case with these changes or no?

**Comment by piranna** [comment]
I've just done your suggestions :-)

File: index.js [hunk 5]
```diff
@@ -396,28 +423,43 @@ SendStream.prototype.isRangeFresh = function isRangeFresh(){
...
+SendStream.prototype.redirect = function redirectDirectory(path) {
   if (listenerCount(this, 'directory') !== 0) {
     this.emit('directory')
     return
   }
 
   if (this.hasTrailingSlash()) {
-    this.error(403)
-    return
+    return this.error(403)
```
Comment (dougwilson) [inline index.js]: This should be returning `undefined`, rather than the event emitter instance.

File: index.js [hunk 6]
```diff
@@ -57,6 +61,32 @@ module.exports.mime = mime
...
+  res.setHeader('X-Content-Type-Options', 'nosniff')
+  res.end(msg)
+}
+
+function redirect(res, loc) {
+  // redirect
+  res.setHeader('Content-Type', 'text/html; charset=UTF-8')
+  res.setHeader('Location', loc)
+
+  loc = escapeHtml(loc)
```
Comment (dougwilson) [inline index.js]: Do not reassign a variable tried into `arguments`, as that results in a v8 deopt. Declare a new variable name, please.

File: index.js [hunk 7]
```diff
@@ -57,6 +61,32 @@ module.exports.mime = mime
 var listenerCount = EventEmitter.listenerCount
   || function(emitter, type){ return emitter.listeners(type).length; };
 
+
+function responseStatus(res, code, msg) {
+  if(msg == null) {
+    msg = statuses[code]
```
Comment (dougwilson) [inline index.js]: Do not reassign a variable tried into `arguments`, as that results in a v8 deopt. Declare a new variable name, please.

File: index.js [hunk 8]
```diff
@@ -57,6 +61,32 @@ module.exports.mime = mime
...
+
+  res.statusCode = code
+  res.setHeader('Content-Length', Buffer.byteLength(msg))
+  res.setHeader('X-Content-Type-Options', 'nosniff')
+  res.end(msg)
+}
+
+function redirect(res, loc) {
+  // redirect
+  res.setHeader('Content-Type', 'text/html; charset=UTF-8')
```
Comment (dougwilson) [inline index.js]: This refactored function is missing the `X-Content-Type-Options` security header the old code was setting. Please add it back :)

File: index.js [hunk 9]
```diff
@@ -396,28 +423,43 @@ SendStream.prototype.isRangeFresh = function isRangeFresh(){
...
+SendStream.prototype.redirectSymbolicLink = function redirectSymbolicLink(path) {
+  var self = this
+
+  fs.readlink(path, function(err, linkString)
+  {
+    if (err) return self.onStatError(err)
+
+    // Get relative paths for all symlinks, also for absolute ones
+    path       = dirname(path)
+    linkString = relative(path, resolve(path, linkString))
```
Comment (dougwilson) [inline index.js]: A symlink can be pointing outside of the `root` set in the constructor. If a redirect occurs like this, the resulting URL is going to be useless is most cases, and in the worse case, the `root` argument is supposed to provide security by boxing all requests into the given root, and the redirect will reveal directory names that could potentially be sensitive. I would expect that the redirect would not work if the symlink pointed outside of the root.

File: index.js (hunk 5 above)
Comment (piranna) [inline index.js]: I should have missed it before. Anyway it doesn't matter if it returns the event-emitter or undefined as far as it's not being used, and in fact I've seen in several places that this is the recomended way to do it. Anyway, I've already fixed it.

File: index.js (hunk 9 above)
Comment (piranna) [inline index.js]: Good point. I think here we have three options:

* Leave it as is, and left to the user the decision if the paths show sensitive info
* Return an error (403 forbidden?)
* Return the actual file content

Maybe we could add an option here to check what behaviour to use...

Comment (dougwilson) [inline index.js]: I would say, as a default, a 403 (option 2) may be the best. Option 3 is not great, because it means this feature is no longer "redirect on symlinks" but instead "maybe redirect on symlinks, maybe not; it depends on what your root is, ultimately". Option 1 is probably not a good option to give.

Comment (piranna) [inline index.js]: So seems we have a winner... :-) I proposed the third one just to make it more transparent. What do you think about adding an option between 1 and 2 (being the 403 the default)?

Comment (dougwilson) [inline index.js]: Just option 2 for now.

**Comment by piranna** [comment]
I've just set 403 error by default if destination is outside files root.

**Comment by piranna** [comment]
Any update on this?

**Comment by dougwilson** [comment]
Hi @piranna, sorry, I have not taken a look since you changed the pull request. Can you at least take a look and fix the failing tests so the pull request passes the test while I take a look at the new changes? Also note that one of the tests is marked as "double callback!" with your changes, so even though that's not causing a failure, you'll need to get that fixed as well.

**Comment by piranna** [comment]
Tests are failling to some unrelated test I can't be able to identify what's the source of the problem ("should return 500 on error when reading file"). Could you be able to take a look so this could be merged?

**Comment by piranna** [comment]
The problem that's giving is about the callback being called twice, probably due to the `nextTick()` call in the test.

**Comment by dougwilson** [comment]
There are still a lot of failing tests and you haven't fixed the double callback, either. In addition, it looks like when you merged, you accidentally reverted code, including a security fix. Please fix all that up and ping me and I'll come back and take a look :)!

**Comment by wesleytodd** [comment]
I am going to close this. It seems like a good *idea* but we would need to have a much more clear and clean change set to ever merge this, which seems unlikely at this point. Feel free to reopen a new PR with *just* the minimal changes required to land the feature if you are willing and able to work on this again in the future.
